README
NAME
    Test::FailWarnings - Add test failures if warnings are caught

VERSION
    version 0.007

SYNOPSIS
    Test file:

        use strict;
        use warnings;
        use Test::More;
        use Test::FailWarnings;

        ok( 1, "first test" );
        ok( 1 + "lkadjaks", "add non-numeric" );

        done_testing;

    Output:

        ok 1 - first test
        not ok 2 - Test::FailWarnings should catch no warnings
        #   Failed test 'Test::FailWarnings should catch no warnings'
        #   at t/bin/main-warn.pl line 7.
        # Warning was 'Argument "lkadjaks" isn't numeric in addition (+) at t/bin/main-warn.pl line 7.'
        ok 3 - add non-numeric
        1..3
        # Looks like you failed 1 test of 3.

DESCRIPTION
    This module hooks $SIG{__WARN__} and converts warnings to Test::More
    "fail()" calls. It is designed to be used with "done_testing", when you
    don't need to know the test count in advance.

    Just as with Test::NoWarnings, this does not catch warnings if other
    things localize $SIG{__WARN__}, as this is designed to catch *unhandled*
    warnings.

USAGE
  Overriding $SIG{__WARN__}
    On "import", $SIG{__WARN__} is replaced with
    "Test::FailWarnings::handler".

        use Test::FailWarnings;  # global

    If you don't want global replacement, require the module instead and
    localize in whatever scope you want.

        require Test::FailWarnings;

        {
            local $SIG{__WARN__} = \&Test::FailWarnings::handler;
            # ... warnings will issue fail() here
        }

    When the handler reports on the source of the warning, it will look past
    any calling packages starting with "Carp" or "warnings" to try to detect
    the real origin of the warning.

  Allowing warnings from dependencies
    If you want to ignore failures from outside your own code, you can set
    $Test::FailWarnings::ALLOW_DEPS to a true value. You can do that on the
    "use" line with "-allow_deps".

        use Test::FailWarnings -allow_deps => 1;

    When true, warnings will only be thrown if they appear to originate from
    a filename matching "qr/^(?:t|xt|lib|blib)/"

  Allowing warnings from specific modules
    If you want to white-list specific modules only, you can add their
    package names to @Test::NoWarnings::ALLOW_FROM. You can do that on the
    "use" line with "-allow_from".

        use Test::FailWarnings -allow_from => [ qw/Annoying::Module/ ];

SEE ALSO
    *   Test::NoWarnings -- catches warnings and reports in an "END" block.
        Not (yet) friendly with "done_testing".

    *   Test::Warnings -- a replacement for Test::NoWarnings that works with
        done_testing

    *   Test::Warn -- test for warnings without triggering failures from
        this modules

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/test-failwarnings/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/test-failwarnings>

      git clone git://github.com/dagolden/test-failwarnings.git

AUTHOR
    David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Test-FailWarnings

0.007     2013-08-30 09:53:47 America/New_York

    [FIXED]

    - In some situations, the source of a local warning can have an
      absolute filename and would not be reported if ALLOW_DEPS were true.
      Now absolute filenames are made relative to the original cwd before
      checking if they are local or not.

    [DOCUMENTED]

    - Fixed synopsis to match new failing test label

0.006     2013-08-22 10:41:11 America/New_York

    [CHANGED]

    - The label of the failing test has been revised for clarity

0.005     2013-05-03 14:03:28 America/New_York

    [DOCUMENTED]

    - Added Test::Warnings to See Also

0.004     2013-05-02 10:50:37 America/New_York

    [FIXED]

    - Bumped Capture::Tiny test prerequisite version

0.003     2013-03-13 13:30:24 America/New_York

    [FIXED]

    - Fixed @ALLOW_FROM initialization to suppress our own warning

0.002     2013-03-10 17:28:53 America/New_York

    [ADDED]

    - Now can optionally allow warnings from dependencies

    - Now can optionally allow warnings from specific modules

    [CHANGED]

    - warnings handler only installed by import(); handler function
      is now available for custom, localized use

0.001     2013-02-25 17:44:33 America/New_York

    - First release



lib/Test/FailWarnings.pm
use 5.008001;
use strict;
use warnings;

package Test::FailWarnings;
# ABSTRACT: Add test failures if warnings are caught
our $VERSION = '0.007'; # VERSION

use Test::More 0.86;
use Cwd qw/getcwd/;
use File::Spec;
use Carp;

our $ALLOW_DEPS = 0;
our @ALLOW_FROM = ();

my $ORIG_DIR = getcwd(); # cache in case handler runs after a chdir

sub import {
    my ( $class, @args ) = @_;
    croak("import arguments must be key/value pairs")
      unless @args % 2 == 0;
    my %opts = @args;
    $ALLOW_DEPS = $opts{'-allow_deps'};
    if ( $opts{'-allow_from'} ) {
        @ALLOW_FROM =
          ref $opts{'-allow_from'} ? @{ $opts{'-allow_from'} } : $opts{'-allow_from'};
    }
    $SIG{__WARN__} = \&handler;
}

sub handler {
    my $msg = shift;
    $msg = '' unless defined $msg;
    chomp $msg;
    my ( $package, $filename, $line ) = _find_source();

    # shortcut if ignoring dependencies and warning did not
    # come from something local
    if ( $ALLOW_DEPS ) {
        $filename = File::Spec->abs2rel( $filename, $ORIG_DIR )
            if File::Spec->file_name_is_absolute( $filename );
        return if $filename !~ /^(?:t|xt|lib|blib)/;
    }

    return if grep { $package eq $_ } @ALLOW_FROM;

    if ( $msg !~ m/at .*? line \d/ ) {
        chomp $msg;
        $msg = "'$msg' at $filename line $line.";
    }
    else {
        $msg = "'$msg'";
    }
    my $builder = Test::More->builder;
    $builder->ok( 0, "Test::FailWarnings should catch no warnings" )
      or $builder->diag("Warning was $msg");
}

sub _find_source {
    my $i = 1;
    while (1) {
        my ( $pkg, $filename, $line ) = caller( $i++ );
        return caller( $i - 2 ) unless defined $pkg;
        next if $pkg =~ /^(?:Carp|warnings)/;
        return ( $pkg, $filename, $line );
    }
}

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding utf-8

=head1 NAME

Test::FailWarnings - Add test failures if warnings are caught

=head1 VERSION

version 0.007

=head1 SYNOPSIS

Test file:

    use strict;
    use warnings;
    use Test::More;
    use Test::FailWarnings;

    ok( 1, "first test" );
    ok( 1 + "lkadjaks", "add non-numeric" );

    done_testing;

Output:

    ok 1 - first test
    not ok 2 - Test::FailWarnings should catch no warnings
    #   Failed test 'Test::FailWarnings should catch no warnings'
    #   at t/bin/main-warn.pl line 7.
    # Warning was 'Argument "lkadjaks" isn't numeric in addition (+) at t/bin/main-warn.pl line 7.'
    ok 3 - add non-numeric
    1..3
    # Looks like you failed 1 test of 3.

=head1 DESCRIPTION

This module hooks C<$SIG{__WARN__}> and converts warnings to L<Test::More>
C<fail()> calls.  It is designed to be used with C<done_testing>, when you
don't need to know the test count in advance.

Just as with L<Test::NoWarnings>, this does not catch warnings if other things
localize C<$SIG{__WARN__}>, as this is designed to catch I<unhandled> warnings.

=for Pod::Coverage handler

=head1 USAGE

=head2 Overriding C<$SIG{__WARN__}>

On C<import>, C<$SIG{__WARN__}> is replaced with
C<Test::FailWarnings::handler>.

    use Test::FailWarnings;  # global

If you don't want global replacement, require the module instead and localize
in whatever scope you want.

    require Test::FailWarnings;

    {
        local $SIG{__WARN__} = \&Test::FailWarnings::handler;
        # ... warnings will issue fail() here
    }

When the handler reports on the source of the warning, it will look past
any calling packages starting with C<Carp> or C<warnings> to try to detect
the real origin of the warning.

=head2 Allowing warnings from dependencies

If you want to ignore failures from outside your own code, you can set
C<$Test::FailWarnings::ALLOW_DEPS> to a true value.  You can
do that on the C<use> line with C<< -allow_deps >>.

    use Test::FailWarnings -allow_deps => 1;

When true, warnings will only be thrown if they appear to originate from a filename
matching C<< qr/^(?:t|xt|lib|blib)/ >>

=head2 Allowing warnings from specific modules

If you want to white-list specific modules only, you can add their package
names to C<@Test::NoWarnings::ALLOW_FROM>.  You can do that on the C<use> line
with C<< -allow_from >>.

    use Test::FailWarnings -allow_from => [ qw/Annoying::Module/ ];

=head1 SEE ALSO

=over 4

=item *

L<Test::NoWarnings> -- catches warnings and reports in an C<END> block.  Not (yet) friendly with C<done_testing>.

=item *

L<Test::Warnings> -- a replacement for Test::NoWarnings that works with done_testing

=item *

L<Test::Warn> -- test for warnings without triggering failures from this modules

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/test-failwarnings/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/test-failwarnings>

  git clone git://github.com/dagolden/test-failwarnings.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


Makefile.PL

use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Add test failures if warnings are caught",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Test-FailWarnings",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Test::FailWarnings",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Cwd" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.86",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => "0.12",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => 0,
    "Test::More" => "0.96",
    "constant" => 0,
    "lib" => 0
  },
  "VERSION" => "0.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



